crates/forge_analyzer/src/ir.rs

```diff
@@ -392,13 +416,185 @@ impl Body {
+                    }
+                }
+                a = pool[a.unwrap()].next;
+            }
+
+            best[v as usize] = sdom[v as usize];
+            idom[v as usize] = uf[v as usize];
+        }
+
+        for i in 1..tick {
+            let v = rdfn[i as usize];
+            while dfn[idom[v as usize] as usize] > dfn[sdom[v as usize] as usize] {
+                idom[v as usize] = idom[idom[v as usize] as usize];
+            }
+        }
+        idom
+    }
+
+    // dfs; helper for semi-nca algo in build_dom_tree()
+    fn dfs(&self, u: usize, tick: &mut u32, dfn: &mut Vec<i32>, rdfn: &mut Vec<i32>,
+           uf: &mut Vec<i32>, outgoing: &mut Vec<Option<usize>>, pool: &Vec<Arc>) {
+        dfn[u] = *tick as i32;
+        rdfn[*tick as usize] = u as i32;
+        *tick += 1;
+
+        let mut a = outgoing[u];
+        while let Some(arc_index) = a {
+            let arc = &pool[arc_index];
+            let v = arc.v;
+            if dfn[v as usize] < 0 {
+                uf[v as usize] = u as i32;
+                self.dfs(v as usize, tick, dfn, rdfn, uf, outgoing, pool);
+            }
+            a = arc.next;
+        }
+    }
+
+    // eval; helper for semi-nca algo in build_dom_tree()
+    fn eval(&self, v: usize, cur: i32, dfn: &Vec<i32>, best: &mut Vec<i32>, uf: &mut Vec<i32>) -> i32 {
+        if dfn[v] <= cur {
+            return v.try_into().unwrap();
+        }
+        let u = uf[v];
+        let r = self.eval(u.try_into().unwrap(), cur, dfn, best, uf);
+        if dfn[best[u as usize] as usize] < dfn[best[v as usize] as usize] {
+            best[v] = best[u as usize];
+        }
+        uf[v] = r;
+        r
+    }
+
+    // returns true if a dominates b; false otherwise
+    // a dominates b if every path from entry -> b must go through a; [entry...a...b]
+    pub(crate) fn dominates(&self, a: BasicBlockId, b: BasicBlockId) -> bool {
+        let dom_tree = self.dominator_tree();
+        let idom = &dom_tree.idom;
+
+        let a = a.0 as i32;
+        let b = b.0 as i32;
+
+        let mut val = b as usize;
+        while idom[val] != -1 {
+            if idom[val] == a {
+                return true;
+            }
+            val = idom[val] as usize;
+        }
+        false
+    }
+    
+    // returns an iterator over the dominance frontier of a basic block
+    pub(crate) fn dominance_frontier(&self, b: BasicBlockId) -> impl Iterator<Item = BasicBlockId> + '_ {
+        let idom = &self.dominator_tree().idom;
+        let mut frontiers: Vec<Vec<BasicBlockId>> = Vec::new();
+        for _ in 0..self.blocks.len() {
+            frontiers.push(Vec::new());
+        }
+
+        // algorithm from https://en.wikipedia.org/wiki/Static_single-assignment_form#Computing_minimal_SSA_using_dominance_frontiers
+        // which references: https://www.cs.tufts.edu/comp/150FP/archive/keith-cooper/dom14.pdf
+        for (id, block) in self.iter_blocks_enumerated() {
```

Comment:
replace "block" with _ since you're not using "block" in the loop.

Reply:
noted - thank you @aliner-wang! I think I have some other places with the same issue, will make those changes as well